Don't default publishDecorations and showWorkspaceLoaded to true on the server

flodiebold opened the pull request:
If the client doesn't specify these explicitly, that very likely means it
doesn't know about them and so we shouldn't send them. In particular, the recent
change to this default caused decorations to be sent to emacs, resulting in a
lot of warning spam.

kjeremy commented:
This may be reasonable in the short term... but long term we really should use the [workspace/configuration](https://microsoft.github.io/language-server-protocol/specification#workspace_configuration) request. `InitializationOptions` is intended to be used more like a list of command line arguments (and iirc may be deprecated in the future).

I started down this path in #806 but haven't had time to finish it (but it should not be that hard).

kjeremy commented:
Note that `workspace/configuration` goes in the opposite direction of most request/responses. The server asks the client what it's configuration should be and the client responds (and of course we have sensible defaults on the server side). This works in tandem with the `didConfigurationChange` notification.

matklad commented:
Agree that publish decorations should be false by default, but I think it makes sense to enable “workspace loaded” notification by default.

Unlike decorations hack, it uses the LSP feature, and should be supported by all the Clients. It is also a very useful troubleshooting tool: if a user is in trouble, asking “have you seen workspace loaded” notification should help us bisect possible causes.

I do think that advance users might want to just disable this notification, but the default is specifically aimed at people who launch rust analyzer for the first time.

DJMcNab commented:
I think decorations should be in the Client Capabilities somehow, I.e we make a 'proper' protocol extension using the right apis in vscode. However, they should be able to be disabled in the config.

flodiebold commented:
> Agree that publish decorations should be false by default, but I think it makes sense to enable “workspace loaded” notification by default.

You're right... I'll change it when I get the time.

flodiebold commented:
Ok, changed it back to default showWorkspaceLoaded to true, and simplified the default handling a bit (so they only have to be in two places instead of three...)

The file had CRLF line endings by the way, that's why it now looks as if the whole file changed. (`?w=1` shows the actual changes)

matklad commented:
bors r+
